### PR TITLE
USB: Add `lsusb` and LibUSBDB

### PR DESCRIPTION
**Pull request by Quaker762**
Also implements some stuff in ProcFS to actually store information about the USB devices attached to the machine.

Relies on #5542 to be merged before this will work.

image

### CODE REVIEW

**Comment by Quaker762** [comment]
Uhh, 9151001 should be dropped when the other PR is merged :^)

File: Kernel/Devices/USB/USBDevice.cpp [hunk 1]
```diff
@@ -16,10 +16,32 @@ static u32 s_next_usb_address = 1; // Next address we hand out to a device once
...
 void USBDevice::create(PortNumber port, DeviceSpeed speed)
 {
-    USBDevice device(port, speed);
-    device.enumerate();
+    auto device = new USBDevice(port, speed);
+    device->enumerate();
+    s_devices.at(static_cast<u8>(port)) = device;
+}
+
+USBDevice* USBDevice::get(PortNumber port)
```
Comment (Quaker762) [inline Kernel/Devices/USB/USBDevice.cpp:28]: Is there a better way of doing this...?

**Comment by Quaker762** [comment]
> I'll kindly ask you to wait a bit until #7509 is merged. If you still want to get this merged ASAP, that's fine too, and I'll try to adapt the changes afterwards :)

Might be good to get this merged as a proof of concept then redo it when we get SysFS merged in tree (the USB stuff is getting stale the longer it waits). We can link up on Discord to get this refactored when SysFS is ready :P

**Comment by Quaker762** [comment]
@supercomputer7 Or, alternatively, how far along is your SysFS stuff? If possible, can I rebase this PR on top of that and start work right away?

**Comment by supercomputer7** [comment]
> @supercomputer7 Or, alternatively, how far along is your SysFS stuff? If possible, can I rebase this PR on top of that and start work right away?

technically yes, the SysFS code is ready and you can add things upon it quite easily :)

File: Userland/Utilities/lsusb.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+#include <AK/JsonObject.h>
+#include <AK/LexicalPath.h>
+#include <AK/String.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/DirIterator.h>
+#include <LibCore/File.h>
+#include <stdio.h>
+#include <unistd.h>
+
+static const char* format = "Device {}: ID {:04x}:{:04x}";
```
Comment (alimpfard) [inline Userland/Utilities/lsusb.cpp]: Please don't do this, I'm pretty sure this circumvents the compiletime format string checker.
it's also used only once, so this is pretty pointless :P

File: Userland/Utilities/lsusb.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+{
+    if (pledge("stdio rpath", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    if (unveil("/proc/bus/usb", "r") < 0) {
+        perror("unveil");
+        return 1;
+    }
```
Comment (alimpfard) [inline Userland/Utilities/lsusb.cpp:31]: Also please lock the veil.
```c++
if (unveil(nullptr, nullptr) < 0) {
    perror("unveil");
    return 1;
}
```

File: Userland/Libraries/LibUSBDB/Database.cpp [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+static T parse_hex(StringView str, size_t count)
+{
+    VERIFY(str.length() >= count);
+
+    T res = 0;
+    for (size_t i = 0; i < count; i++)
+        res = (res << 4) + parse_hex_digit(str[i]);
+
+    return res;
+}
```
Comment (alimpfard) [inline Userland/Libraries/LibUSBDB/Database.cpp:111]: We have half a million of these in the codebase, please use `AK::StringUtils::convert_from_uint_from_hex<T>(StringView)` instead.
`count` is also pointless to take, it's the size of T.

Comment (Quaker762) [reply Userland/Libraries/LibUSBDB/Database.cpp:111]: Copy and paste from elsewhere in the source has, once again, failed me lol.

**Pull request merged**